lib/src/ui/puzzle/puzzle_history_screen.dart

```diff
@@ -0,0 +1,263 @@
+  }
+
+  Widget _buildIos(BuildContext context) {
+    return CupertinoPageScaffold(
+      navigationBar: CupertinoNavigationBar(
+        middle: Text(context.l10n.puzzleHistory),
+      ),
+      child: _Body(historyList),
+    );
+  }
+
+  Widget _buildAndroid(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(title: Text(context.l10n.puzzleHistory)),
+      body: _Body(historyList),
+    );
+  }
+}
+
+class _Body extends ConsumerStatefulWidget {
+  const _Body(this.historyList);
+  final IList<PuzzleHistoryEntry> historyList;
+  @override
+  ConsumerState<_Body> createState() => _BodyState();
+}
+
+class _BodyState extends ConsumerState<_Body> {
+  final ScrollController _scrollController = ScrollController();
+  List<PuzzleHistoryEntry> _historyList = [];
+  bool _isLoading = false;
+  bool _hasMore = true;
+
+  @override
+  void initState() {
+    super.initState();
+    _historyList.addAll(widget.historyList);
+    _scrollController.addListener(_scrollListener);
+  }
+
+  @override
+  void dispose() {
+    _scrollController.removeListener(_scrollListener);
+    _scrollController.dispose();
+    super.dispose();
+  }
+
+  void _scrollListener() {
+    if (_scrollController.position.pixels ==
+            _scrollController.position.maxScrollExtent &&
+        _historyList.length < 500) {
+      _loadMore();
+    }
+  }
+
+  Future<void> _loadMore() async {
+    if (_isLoading) return;
+    if (!_hasMore) return;
+    setState(() {
+      _isLoading = true;
+    });
+    final newList = await ref
+        .read(puzzleActivityProvider(50, _historyList.last.date).future);
+
+    if (mounted) {
+      if (newList.isEmpty) {
+        setState(() {
+          _isLoading = false;
+          _hasMore = false;
+        });
+        return;
+      }
+      setState(() {
+        _isLoading = false;
+        _historyList.addAll(newList);
+        _historyList = _historyList.toSet().toList();
+      });
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final crossAxisCount =
+        MediaQuery.of(context).size.width > kTabletThreshold ? 4 : 2;
+    final boardWidth = MediaQuery.of(context).size.width / crossAxisCount;
+    return ListView.builder(
+      controller: _scrollController,
+      itemCount: _historyList.length ~/ crossAxisCount + (_isLoading ? 1 : 0),
+      itemBuilder: (context, index) {
+        if (_isLoading && index == _historyList.length ~/ crossAxisCount) {
+          return const CenterLoadingIndicator();
+        }
+        final rowStartIndex = index * crossAxisCount;
+        final rowEndIndex =
+            min(rowStartIndex + crossAxisCount, _historyList.length);
+        return Row(
+          children: _historyList
+              .getRange(index * crossAxisCount, rowEndIndex)
+              .map(
+                (e) => _HistoryBoard(e, boardWidth),
+              )
+              .toList(),
+        );
+      },
+    );
+  }
+}
+
+class _HistoryBoard extends ConsumerWidget {
```

Comment:
This widget is used instead of `PuzzleHistoryBoards` to support lazy loading of `ListView.builder`